### PR TITLE
add memory usage limit option to Engine

### PR DESCRIPTION
**Pull request by 0xFireball**

### CODE REVIEW

File: Jint/Engine.cs [hunk 1]
```diff
@@ -370,6 +383,17 @@ public Completion ExecuteStatement(Statement statement)
                 throw new TimeoutException();
             }
 
+#if NETCOREAPP2_0
+            if (Options._MemoryLimit > 0)
+            {
+                var memoryUsage = GC.GetAllocatedBytesForCurrentThread() - _initialMemoryUsage;
+                if (memoryUsage > Options._MemoryLimit)
+                {
+                    throw new MemoryLimitExceededException();
```
Comment (ayende) [inline Jint/Engine.cs]: I would suggest adding some details here. Script has allocated X but is limited to Y

File: Jint/Runtime/MemoryLimitExceededException.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+
+namespace Jint.Runtime
+{
+    public class MemoryLimitExceededException : Exception
+    {
+        public MemoryLimitExceededException() : base("The memory limit has been exceeded.")
```
Comment (ayende) [inline Jint/Runtime/MemoryLimitExceededException.cs]: I would recommend using standard exception ctors, instead.

File: Jint/Engine.cs [hunk 3]
```diff
@@ -372,6 +385,17 @@ public Completion ExecuteStatement(Statement statement)
                 throw new TimeoutException();
             }
 
+#if NETCOREAPP2_0
+            if (Options._MemoryLimit > 0)
+            {
+                var memoryUsage = GC.GetAllocatedBytesForCurrentThread() - _initialMemoryUsage;
```
Comment (pranavkm) [inline Jint/Engine.cs]: Might be a bit more cumbersome, but you could avoid cross-compiling by reflecting on the method. This should make it work in all platforms that support the API.

`Func<int> GetAllocatedBytesForCurrentThread = _methodInfo != null ? Delegate.CreateDelegate() : long.MaxValue;`

File: Jint/Engine.cs [hunk 4]
```diff
@@ -385,16 +400,14 @@ public Completion ExecuteStatement(Statement statement)
                 throw new TimeoutException();
             }
 
-#if NETCOREAPP2_0
             if (Options._MemoryLimit > 0)
             {
-                var memoryUsage = GC.GetAllocatedBytesForCurrentThread() - _initialMemoryUsage;
+                var memoryUsage = GetAllocatedBytesForCurrentThread() - _initialMemoryUsage;
                 if (memoryUsage > Options._MemoryLimit)
```
Comment (ayende) [inline Jint/Engine.cs]: This will _always_ be triggered if the method is not available.

File: Jint/Engine.cs [hunk 5]
```diff
@@ -76,6 +74,20 @@ public class Engine
...
+        {
+            var methodInfo = typeof(GC).GetMethod("GetAllocatedBytesForCurrentThread");
+
+            if (methodInfo != null)
+            {
+                GetAllocatedBytesForCurrentThread =  (Func<long>)Delegate.CreateDelegate(typeof(Func<long>), null, methodInfo);
+            }
+            else
+            {
+                GetAllocatedBytesForCurrentThread = () => long.MaxValue;
```
Comment (ayende) [inline Jint/Engine.cs]: Why not returning 0 here?

File: Jint/Engine.cs [hunk 6]
```diff
@@ -247,6 +258,8 @@ internal ReferencePool ReferencePool
         }
         #endregion
 
+        private static Func<long> GetAllocatedBytesForCurrentThread;
```
Comment (pranavkm) [inline Jint/Engine.cs]: `static readonly`

File: Jint/Engine.cs [hunk 7]
```diff
@@ -384,6 +411,22 @@ public Completion ExecuteStatement(Statement statement)
...
+                {
+                    var memoryUsage = GetAllocatedBytesForCurrentThread() - _initialMemoryUsage;
+                    if (memoryUsage > Options._MemoryLimit)
+                    {
+                        throw new MemoryLimitExceededException($"Script has allocated {memoryUsage} but is limited to {Options._MemoryLimit}");
+                    }
+                }
+                else
+                {
+                    throw new PlatformNotSupportedException("The current platform doesn't support MemoryLimit.");
```
Comment (pranavkm) [inline Jint/Engine.cs]: 👏

**Pull request merged**